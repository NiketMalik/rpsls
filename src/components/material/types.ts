import type { MATERIAL_TYPES } from "constants/material";

export interface MaterialProps {
  type?: MATERIAL_TYPES;
  radius?: number;
  isDisabled?: boolean;
  isHighlited?: boolean;
  onClick?: () => void;
}
