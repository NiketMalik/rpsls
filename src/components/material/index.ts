export { Material } from "./Material";
export { DEFAULT_RADIUS } from "./constants";
export type { MaterialProps } from "./types";
