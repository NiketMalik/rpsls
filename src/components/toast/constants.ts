export enum TOAST_ACCENTS {
  "default" = "default",
  "primary" = "primary",
  "secondary" = "secondary",
  "info" = "info",
  "success" = "success",
  "danger" = "danger",
  "warning" = "warning",
}
