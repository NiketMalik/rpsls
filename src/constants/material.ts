export enum MATERIAL_TYPES {
  ROCK = "ROCK",
  PAPER = "PAPER",
  SCISSOR = "SCISSOR",
  LIZARD = "LIZARD",
  SPOCK = "SPOCK",
}
