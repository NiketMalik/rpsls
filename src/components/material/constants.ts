import { MATERIAL_TYPES } from "constants/material";

import rockIcon from "static/images/material-rock.svg";
import paperIcon from "static/images/material-paper.svg";
import scissorIcon from "static/images/material-scissor.svg";
import lizardIcon from "static/images/material-lizard.svg";
import spockIcon from "static/images/material-spock.svg";

export const MATERIAL_ICONS_MAP = {
  [MATERIAL_TYPES.ROCK]: rockIcon,
  [MATERIAL_TYPES.PAPER]: paperIcon,
  [MATERIAL_TYPES.SCISSOR]: scissorIcon,
  [MATERIAL_TYPES.LIZARD]: lizardIcon,
  [MATERIAL_TYPES.SPOCK]: spockIcon,
};

export const DEFAULT_RADIUS = 60;
