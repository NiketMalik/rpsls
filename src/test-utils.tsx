import { FC, ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { THEME } from "constants/theme";

import { StoreProvider } from "context/storeContext";
import { ThemeProvider } from "@emotion/react";

const WithGlobalProviders: FC = ({ children }) => {
  return (
    <StoreProvider>
      <ThemeProvider theme={THEME}>{children}</ThemeProvider>
    </StoreProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithGlobalProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
