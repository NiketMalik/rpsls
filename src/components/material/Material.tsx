import { memo } from "react";

import type { MaterialProps } from "./types";

import { DEFAULT_RADIUS, MATERIAL_ICONS_MAP } from "./constants";

import { MaterialContainer, MaterialImage, MaterialHighlight } from "./styles";

export const Material = memo(
  ({ type, radius, isDisabled, isHighlited, onClick }: MaterialProps) => {
    const materialIconUrl = type && MATERIAL_ICONS_MAP[type];

    return (
      <MaterialContainer
        aria-label={type}
        material={type}
        radius={radius || DEFAULT_RADIUS}
        onClick={!isDisabled ? onClick : undefined}
        data-disabled={isDisabled || !materialIconUrl}
        data-is-highlited={!!isHighlited}
        data-testid="material"
      >
        {materialIconUrl && (
          <MaterialImage
            src={materialIconUrl}
            alt={type}
            data-testid="material-image"
          />
        )}
        <MaterialHighlight data-testid="material-highlight" />
      </MaterialContainer>
    );
  },
);
