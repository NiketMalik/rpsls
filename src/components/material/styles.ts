import styled from "@emotion/styled/macro";

import { MATERIAL_TYPES } from "constants/material";

import { getLinearGradient } from "utils/styles";

interface MaterialContainerProps {
  material?: MATERIAL_TYPES;
  radius: number;
}

export const MaterialHighlight = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;

  &::before,
  &::after {
    content: "";
    position: absolute;
    border-radius: 50%;
    background: ${(props) => props.theme.gradients.background.start};
    transition: transform 0.25s ease-in-out;
    pointer-events: none;
  }

  &::before {
    width: 220%;
    height: 220%;
    top: -60%;
    left: -60%;

    opacity: 0.1;
  }

  &::after {
    width: 160%;
    height: 160%;
    top: -30%;
    left: -30%;

    opacity: 0.15;
  }
`;

export const MaterialContainer = styled.button<MaterialContainerProps>`
  position: relative;
  width: ${(props) => props.radius * 2}px;
  height: ${(props) => props.radius * 2}px;
  min-width: ${(props) => props.radius * 2}px;
  min-height: ${(props) => props.radius * 2}px;

  &::before,
  &::after {
    content: "";
    position: absolute;
    border-radius: 50%;
    z-index: 1;
  }

  &::before {
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;

    ${(props) =>
      getLinearGradient(
        props.theme.gradients.material[props.material || "DEFAULT"],
      )}
    opacity: ${(props) => (props.material ? "1" : "0.5")};
    box-shadow: ${(props) =>
      props.material ? "inset 0 -5px rgba(0, 0, 0, 0.15)" : "none"};
  }

  &::after {
    display: ${(props) => (props.material ? "block" : "none")};
    top: 50%;
    left: 50%;
    width: 80%;
    height: 80%;
    transform: translate(-50%, -50%);

    background: white;
    box-shadow: inset 0 7px rgba(0, 0, 0, 0.15);
  }

  &:not([disabled]):active {
    transform: scale(0.95);
  }

  &[data-is-highlited="true"] {
    margin: 30%;

    ${MaterialHighlight} {
      &::before,
      &::after {
        transform: scale(1);
      }
    }
  }

  &[data-is-highlited="false"] {
    ${MaterialHighlight} {
      &::before,
      &::after {
        transform: scale(0);
      }
    }
  }
`;

export const MaterialImage = styled.img`
  position: absolute;
  top: 50%;
  left: 50%;
  width: 45%;
  height: 45%;
  transform: translate(-50%, -50%);
  z-index: 2;
`;
