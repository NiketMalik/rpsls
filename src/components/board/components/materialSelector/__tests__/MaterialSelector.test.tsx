import React from "react";

import { MATERIAL_TYPES } from "constants/material";
import { fireEvent, render, screen } from "test-utils";

import { MaterialSelector } from "../index";

describe("MaterialSelector", () => {
  it("renders without error", () => {
    const { container } = render(<MaterialSelector radius={100} />);
    expect(container).toMatchSnapshot();
  });

  describe("actions", () => {
    it("onSelect", () => {
      const mockOnSelect = jest.fn();
      render(<MaterialSelector radius={100} onSelect={mockOnSelect} />);

      fireEvent.click(screen.getAllByTestId("material")[0]);
      expect(mockOnSelect).toBeCalledWith(MATERIAL_TYPES.ROCK);
    });
  });
});
