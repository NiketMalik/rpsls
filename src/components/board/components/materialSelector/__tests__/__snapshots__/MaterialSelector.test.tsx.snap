// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaterialSelector renders without error 1`] = `
<div>
  <div
    class="d-flex justify-content-center m-2 m-md-4 col-md-4"
    style="opacity: 0; transform: scale(0) translateZ(0);"
  >
    <button
      aria-label="ROCK"
      class="css-3gmqhb-MaterialContainer ec4hnvt1"
      data-disabled="false"
      data-is-highlited="false"
      data-testid="material"
      radius="100"
    >
      <img
        alt="ROCK"
        class="css-stx9by-MaterialImage ec4hnvt0"
        data-testid="material-image"
        src="material-rock.svg"
      />
      <div
        class="css-1nc6hb2-MaterialHighlight ec4hnvt2"
        data-testid="material-highlight"
      />
    </button>
  </div>
  <div
    class="d-flex justify-content-center m-2 m-md-4 col-md-4"
    style="opacity: 0; transform: scale(0) translateZ(0);"
  >
    <button
      aria-label="PAPER"
      class="css-1hp7zwx-MaterialContainer ec4hnvt1"
      data-disabled="false"
      data-is-highlited="false"
      data-testid="material"
      radius="100"
    >
      <img
        alt="PAPER"
        class="css-stx9by-MaterialImage ec4hnvt0"
        data-testid="material-image"
        src="material-paper.svg"
      />
      <div
        class="css-1nc6hb2-MaterialHighlight ec4hnvt2"
        data-testid="material-highlight"
      />
    </button>
  </div>
  <div
    class="d-flex justify-content-center m-2 m-md-4 col-md-4"
    style="opacity: 0; transform: scale(0) translateZ(0);"
  >
    <button
      aria-label="SCISSOR"
      class="css-1l31af5-MaterialContainer ec4hnvt1"
      data-disabled="false"
      data-is-highlited="false"
      data-testid="material"
      radius="100"
    >
      <img
        alt="SCISSOR"
        class="css-stx9by-MaterialImage ec4hnvt0"
        data-testid="material-image"
        src="material-scissor.svg"
      />
      <div
        class="css-1nc6hb2-MaterialHighlight ec4hnvt2"
        data-testid="material-highlight"
      />
    </button>
  </div>
  <div
    class="d-flex justify-content-center m-2 m-md-4 col-md-4"
    style="opacity: 0; transform: scale(0) translateZ(0);"
  >
    <button
      aria-label="LIZARD"
      class="css-1fa8azu-MaterialContainer ec4hnvt1"
      data-disabled="false"
      data-is-highlited="false"
      data-testid="material"
      radius="100"
    >
      <img
        alt="LIZARD"
        class="css-stx9by-MaterialImage ec4hnvt0"
        data-testid="material-image"
        src="material-lizard.svg"
      />
      <div
        class="css-1nc6hb2-MaterialHighlight ec4hnvt2"
        data-testid="material-highlight"
      />
    </button>
  </div>
  <div
    class="d-flex justify-content-center m-2 m-md-4 col-md-4"
    style="opacity: 0; transform: scale(0) translateZ(0);"
  >
    <button
      aria-label="SPOCK"
      class="css-bj4oie-MaterialContainer ec4hnvt1"
      data-disabled="false"
      data-is-highlited="false"
      data-testid="material"
      radius="100"
    >
      <img
        alt="SPOCK"
        class="css-stx9by-MaterialImage ec4hnvt0"
        data-testid="material-image"
        src="material-spock.svg"
      />
      <div
        class="css-1nc6hb2-MaterialHighlight ec4hnvt2"
        data-testid="material-highlight"
      />
    </button>
  </div>
</div>
`;
